### ISSUE TITLE
Weld problems with AbstractDicomService#getSOPClasses() final method

### ISSUE THREAD

**Issue by craigtmoore** [dcm4che/dcm4che#1010]
@gunterze we're running into some problems loading this class in JBoss EAP7.3 with weld because of the final getSOPClasses() method seen here: https://github.com/dcm4che/dcm4che/blob/3543c9dd8d3f101ba4f9b6099b02585fdd0f0c5b/dcm4che-net/src/main/java/org/dcm4che3/net/service/AbstractDicomService.java#L62

Here's an example of the stacktrace, with the WeldException:

```
org.jboss.msc.service.StartException in service jboss.deployment.unit."agility.ear".WeldStartService: Failed to start service
    at org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1731)
    at org.jboss.msc.service.ServiceControllerImpl$ControllerTask.run(ServiceControllerImpl.java:1559)
    at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
    at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
    at java.lang.Thread.run(Thread.java:748)
Caused by: org.jboss.weld.exceptions.WeldException: WELD-001524: Unable to load proxy class for bean Managed Bean [class org.dcm4chee.archive.store.scp.CStoreSCP] with qualifiers [@Any @Default] with class class org.dcm4chee.archive.store.scp.CStoreSCP
    at org.jboss.weld.bean.proxy.ProxyFactory.getProxyClass(ProxyFactory.java:378)
    at org.jboss.weld.injection.producer.SubclassedComponentInstantiator.createEnhancedSubclass(SubclassedComponentInstantiator.java:113)
    at org.jboss.weld.injection.producer.SubclassedComponentInstantiator.initEnhancedSubclass(SubclassedComponentInstantiator.java:86)
    at org.jboss.weld.injection.producer.SubclassedComponentInstantiator.<init>(SubclassedComponentInstantiator.java:79)
    at org.jboss.weld.injection.producer.SubclassedComponentInstantiator.forInterceptedDecoratedBean(SubclassedComponentInstantiator.java:63)
    at org.jboss.weld.injection.producer.BeanInjectionTarget.initializeAfterBeanDiscovery(BeanInjectionTarget.java:121)
    at org.jboss.weld.injection.producer.InjectionTargetInitializationContext.initialize(InjectionTargetInitializationContext.java:42)
    at org.jboss.weld.injection.producer.InjectionTargetService.initialize(InjectionTargetService.java:63)
    at org.jboss.weld.bootstrap.WeldStartup.deployBeans(WeldStartup.java:481)
    at org.jboss.weld.bootstrap.WeldBootstrap.deployBeans(WeldBootstrap.java:86)
    at org.jboss.as.weld.WeldStartService.start(WeldStartService.java:97)
    at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1739)
    at org.jboss.msc.service.ServiceControllerImpl$StartTask.execute(ServiceControllerImpl.java:1701)
    ... 6 more
Caused by: java.lang.VerifyError: class org.dcm4chee.archive.store.scp.DicomService$1944916883$Proxy$_$$_WeldSubclass overrides final method getSOPClasses.()[Ljava/lang/String;
    at java.lang.ClassLoader.defineClass1(Native Method)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
    at java.lang.ClassLoader.defineClass(ClassLoader.java:832)
    at org.jboss.modules.ModuleClassLoader.defineClassInternal(ModuleClassLoader.java:473)
    at org.jboss.modules.ModuleClassLoader.defineClassInternal(ModuleClassLoader.java:480)
    at org.jboss.modules.ClassDefiner.defineClass(ClassDefiner.java:144)
    at org.jboss.as.weld.services.bootstrap.ProxyServicesImpl.defineClass(ProxyServicesImpl.java:204)
    at org.jboss.weld.util.bytecode.ClassFileUtils.toClass(ClassFileUtils.java:132)
    at org.jboss.weld.bean.proxy.ProxyFactory.createProxyClass(ProxyFactory.java:479)
    at org.jboss.weld.bean.proxy.ProxyFactory.getProxyClass(ProxyFactory.java:370)
    ... 18 more
```

As you can see the org.jboss.weld.bean.proxy.ProxyFactory.createProxyClass fails because this method is final. I was wondering if you had any ideas how we might approach this. Could we remove the 'final' modifier?

**Comment by gunterze** [comment]
And you inject them by
```
    @Inject
    private Instance<DicomService> dicomServices;
```
in which type of managed object?
What does Red Hat support says about?

**Comment by craigtmoore** [comment]
Hi Gunter, so the CStoreSCP class is injected by the ArchiveServiceImpl class (I think it was written when you were still at AGFA). 

```
  42   │ /**
  43   │  * @author Gunter Zeilinger <gunterze@gmail.com>
  44   │  *
  45   │  */
  46   │ @ApplicationScoped
  47   │ @Typed(DicomService.class)
  48   │ public class CStoreSCP extends BasicCStoreSCP {
```

Note the `@Typed(DicomService.class)` annotation. The CStoreSCP has the following heirachy
![CStoreSCP Hierarchy](https://user-images.githubusercontent.com/9006450/129598789-4871249a-95e6-424c-9853-68dec3980d9e.png)

The class is injected by the ArchiveServiceImpl class

```
  69   │ /**
  70   │  * @author Gunter Zeilinger <gunterze@gmail.com>
  71   │  */
  72   │ @Singleton
  73   │ @Startup
  74   │ public class ArchiveServiceImpl implements ArchiveService {
...
 104   │     @Inject
 105   │     private Instance<DicomService> dicomServices;
```

As you can see on line 105 it injects the dicomServices value which it then uses during the init method:

```
 153   │     @PostConstruct
 154   │     public void init() {
...
 165   │             for (DicomService service : dicomServices) {
 166   │                 serviceRegistry.addDicomService(service);
 167   │             }
 168
```

See line 166, the @DicomService class gets added to the serviceRegistry. I'm not exactly sure if things go wrong during the Injection or when loading the service into the serviceRegistry. Likely its a problem with Weld intercepting the injection to add its own Proxy class.

**Comment by gunterze** [comment]
Can you reproduce the failure also on deploying on Wildfly . We never got it, currently using Wildfly 24.0.1.Final including
```
 find /opt/wildfly/modules/ -name "*weld*"
/opt/wildfly/modules/system/layers/base/org/jboss/as/jsf-injection/main/weld-jsf-3.1.7.SP1.jar
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld/beanvalidation/main/wildfly-weld-bean-validation-24.0.1.Final.jar
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld/webservices/main/wildfly-weld-webservices-24.0.1.Final.jar
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld/main/wildfly-weld-24.0.1.Final.jar
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld/common/main/wildfly-weld-common-24.0.1.Final.jar
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld/transactions/main/wildfly-weld-transactions-24.0.1.Final.jar
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld/jpa/main/wildfly-weld-jpa-24.0.1.Final.jar
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld/ejb/main/wildfly-weld-ejb-24.0.1.Final.jar
/opt/wildfly/modules/system/layers/base/org/jboss/as/weld/spi/main/wildfly-weld-spi-24.0.1.Final.jar
/opt/wildfly/modules/system/layers/base/org/jboss/weld
/opt/wildfly/modules/system/layers/base/org/jboss/weld/core/main/weld-jta-3.1.7.SP1.jar
/opt/wildfly/modules/system/layers/base/org/jboss/weld/core/main/weld-core-impl-3.1.7.SP1.jar
/opt/wildfly/modules/system/layers/base/org/jboss/weld/core/main/weld-web-3.1.7.SP1.jar
/opt/wildfly/modules/system/layers/base/org/jboss/weld/core/main/weld-ejb-3.1.7.SP1.jar
/opt/wildfly/modules/system/layers/base/org/jboss/weld/api/main/weld-api-3.1.SP4.jar
/opt/wildfly/modules/system/layers/base/org/jboss/weld/spi/main/weld-spi-3.1.SP4.jar
/opt/wildfly/modules/system/layers/base/org/jboss/weld/probe/main/weld-probe-core-3.1.7.SP1.jar
```
?

**Comment by craigtmoore** [comment]
It would be rather difficult for me to port our code base to WildFly. We've
just spent the last 8 months upgrading from JBoss6 to JBoss7.

**Comment by gunterze** [comment]
Which WELD version is  JBoss EAP7.3 using? What's the answer of Redhat support?

**Comment by craigtmoore** [comment]
Here's what I see in the redhat runtime artifacts pom.xml file for EAP 7.3.8.GA (what we're currently targeting)
* https://maven.repository.redhat.com/ga/org/jboss/bom/eap-runtime-artifacts/7.3.8.GA/eap-runtime-artifacts-7.3.8.GA.pom

```
<version.org.jboss.weld.weld-api>3.1.0.SP3-redhat-00001</version.org.jboss.weld.weld-api>
<version.org.jboss.weld.weld-spi>3.1.0.SP3-redhat-00001</version.org.jboss.weld.weld-spi>
<version.org.jboss.weld.weld-core-impl>3.1.6.Final-redhat-00001</version.org.jboss.weld.weld-core-impl>
<version.org.jboss.weld.module>3.1.6.Final-redhat-00001</version.org.jboss.weld.module>
<version.org.jboss.weld.probe>3.1.6.Final-redhat-00001</version.org.jboss.weld.probe>
```

**Comment by craigtmoore** [comment]
Just as a followup, the issue has gone away, likely this was just some bug that got fixed between 7.3.5 and 7.3.8.
